[MDEV-14978] Client programs to use $MARIADB_HOST consistently

oceanli-hub opened the pull request:
## Description
Note: mysql_upgrade already uses environment variable as default. This feature was not added in for mysql_plugin and mysqltest 

Only `mysql` client program was using $MYSQL_HOST as the default host.
Add the same feature in most other client programs but using $MARIADB_HOST instead.

## How can this PR be tested?

Set up environment variable `export MARIADB_HOST=your_test_server`

Run each client program with the following commands:

`client/mysqlslap -h another_test_server -P 3306 -u admin -p`

`client/mysqlslap  -P 3306 -u admin -p`

Verify that $MARIADB_HOST is used when `-h` is not specified. Also verify $MARIADB_HOST does not override command line argument.

## Basing the PR against the correct MariaDB version
- [X] *This is a new feature and the PR is based against the latest MariaDB development branch*

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

ottok commented:
This is an improved take on #1067. Please review and merge, thanks!

grooverdan commented on client/mysqladmin.cc:
```diff
@@ -245,7 +245,7 @@ static const char *load_default_groups[]=
 get_one_option(const struct my_option *opt, const char *argument, const char *filename)
 {
-
+  
```
please don't add a the whitespace. Not sure of the purpose of adding a new line to the beginning of the function, I can't see this in the coding standard. So please remove line unless a reason for inclusion can be provided.

oceanli-hub replied:
The white space is already there on line 248 in https://github.com/MariaDB/server/blob/11.0/client/mysqladmin.cc. I am removed it and added it back because it was originally there. I am not sure why it is showing I added a new whitespace

grooverdan commented on client/mysqladmin.cc:
```diff
@@ -373,6 +373,11 @@ int main(int argc,char *argv[])
+  
+  tmp= (char *) getenv("MYSQL_HOST");
+  if (tmp && !host)
+    host= my_strdup(PSI_NOT_INSTRUMENTED, tmp, MYF(MY_WME));
```
why are you doing a strdup? Is an assignment enough? This memory isn't `free`d so if a `host= tmp` works can we use that.

oceanli-hub replied:
This was copy and pasted from mysql.cc https://github.com/MariaDB/server/blob/11.0/client/mysql.cc#L2135-L2137. I will try your suggestion and see if it works or free up the memory

grooverdan commented on client/mysqladmin.cc:
```diff
@@ -373,6 +373,11 @@ int main(int argc,char *argv[])
   char **commands, **save_argv, **temp_argv;
+  char *tmp;
+  
+  tmp= (char *) getenv("MYSQL_HOST");
```
As this is new functionally, lets start with with `MARIADB_HOST` and add the same compatibility to the mysql.cc (doesn't have to be this PR).

grooverdan replied:
I'm unsure why this is casting using `(char *)`. I looks like getenv returns a `char *`.

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,64 @@
+--let MYSQLDUMP = $MYSQL_BINDIR/client//mariadb-dump
+--let MYSQLIMPORT = $MYSQL_BINDIR/client//mariadb-import
+--let MYSQLSHOW = $MYSQL_BINDIR/client//mariadb-show
+--let MYSQLSLAP = $MYSQL_BINDIR/client//mariadb-slap
```
These are all set in mysql-test/mariadb-test-run.pl, though usually as with underscore like `$MYSQL_CHECK`.

oceanli-hub replied:
The environment variables for the client programs have a default options file argument in them. The options file was  stopping me from testing the client programs

LinuxJedi replied:
In that case, they should probably be things like  `MARIABINLOG`, etc...

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,64 @@
+--exec $MYSQLSLAP $options > /dev/null 2>&1
+
+# Clean up
+--echo Done
```
cleanup should remove `$MYSQL_TMP_DIR/pet` and `tmp.sql`

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,64 @@
+--let MYSQLSLAP = $MYSQL_BINDIR/client//mariadb-slap
+
+# Creating a database for the client programs
+CREATE DATABASE env_variable;
```
the default `test` database should be sufficient so no need to create one.

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,64 @@
+
+--echo ****************
+--echo # Setting MYSQL_HOST environment variable 
+--let MYSQL_HOST=nonexistent-sever
```
spelling "server"

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,64 @@
+--exec $MYSQLIMPORT $options env_variable --local $MYSQL_TMP_DIR/pet > /dev/null 2>&1
+--exec $MYSQLSHOW $options env_variable > /dev/null 2>&1
+--exec $MYSQLSLAP $options > /dev/null 2>&1
+
```
Are there positive test examples using ~MYSQL_HOST~ `MARIADB_HOST` maybe with `--protocol tcp`?

grooverdan reviewed (changes requested):
If you're able to update the man pages for these client utilities too that would be appreciated.

oceanli-hub commented:
Hi @grooverdan I updated my pull requested based off your feedback. I also updated the help function to include $MARIADB_HOST. In mysqlbinlog and mysqldump the memory is freed at the end during clean up so I was getting SIGSEV error when trying to use assignment. Assignments work for all the other client programs. 

For mysql client program would you like me to change it to $MARIADB_HOST in this pr also? 

Thanks for the review!

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,76 @@
+# Clean up
+--echo Done
+DROP TABLE pet;
+--exec rm $MYSQL_TMP_DIR/tmp.sql
```
`--remove_file` for windows compatibility

grooverdan commented on client/mysqlcheck.c:
```diff
@@ -437,6 +437,9 @@ static int get_options(int *argc, char ***argv)
   DBUG_ENTER("get_options");
 
+  if (current_host == NULL)
+    current_host = getenv("MARIADB_HOST");
```
Style fix (applies to all other code too),

`current_host= getenv...`

No space between var and `=`.

grooverdan commented on mysql-test/suite/client/client-env-variable.test:
```diff
@@ -0,0 +1,76 @@
+EOF
+
+# Options for client program
+--let $options = --user=root --port=16000 --socket=$MYSQL_TMP_DIR/mysqld.1.sock
```
`port` and `socket` aren't fixed.

`$MASTER_MYPORT` and `$MASTER_MYSOCK` (from mysql-test/include/default_my.cnf - no include required - its a default).

Or, if you create a `client-env-variable.cnf` like:
```
!include include/default_my.cnf

[ENV]
MARIADB_HOST=localhost
```

This is a true environment variable.

Then `$options` to be `--port=$MASTER_MYPORT` (opmit socket so it doesn't get used) and use a positive test (no more `--error`) to ensure the connection is there.

oceanli-hub replied:
> no more --error
I am not sure if you want to remove the negative test case with `--error`. I added the positive test case but kept the negative test case in the new revision.

grooverdan reviewed (commented):
> For mysql client program would you like me to change it to $MARIADB_HOST in this pr also?

Sure, but in this case `$MARIADB_HOST` will be preferred over `$MYSQL_HOST` so:
```
if (!(tmp= getenv("MARIADB_HOST"))
  tmp= getenv("MYSQL_HOST")
...
```

oceanli-hub commented:
I pushed a new revision based off feedback. `$MARIADB_HOST` is added the mysql client program. For the MTR test, I created  `client-env-variable.cnf` and added positive tests to show the connection is there.

LinuxJedi commented on client/mysqladmin.cc:
```diff
@@ -374,6 +373,9 @@ int main(int argc,char *argv[])
   MYSQL mysql;
   char **commands, **save_argv, **temp_argv;
 
+  if (host==NULL)
```
Minor style thing, spaces around `==`.

LinuxJedi commented on client/mysql.cc:
```diff
@@ -1704,8 +1704,8 @@ static struct my_option my_long_options[] =
    &opt_nobeep, 0, GET_BOOL, NO_ARG, 0, 0, 0, 0, 0, 0},
-  {"host", 'h', "Connect to host.", &current_host,
-   &current_host, 0, GET_STR_ALLOC, REQUIRED_ARG, 0, 0, 0, 0, 0, 0},
+  {"host", 'h', "Connect to host. If host is not specified, connect to localhost by default or $MARIADB_HOST",
```
I suspect it would read better if this said something along the lines of "Defaults in the following order: environment variables MARIADB_HOST, MYSQL_HOST, and then localhost."

LinuxJedi replied:
Also, as a style thing, line is longer than 80 chars.

LinuxJedi reviewed (changes requested):
Looking good, one more final cleanup pass and I think we are there.

oceanli-hub commented:
@LinuxJedi thanks for the feedback, I pushed a new revision based off your comments. Please review again.

LinuxJedi reviewed (approved):
I think that is the bases covered now, many thanks.

grooverdan reviewed (approved):
looks good. Ready for 11.6 testing.

optional improvement of manual pages welcome.

oceanli-hub commented:
@grooverdan @LinuxJedi I rebased the commit onto 11.6 branch. The failing buildbots appear to be just flaky MTR tests. Can you help re-run those?

oceanli-hub commented:
@grooverdan @LinuxJedi I rebased the commit onto the main branch. Please let me know if you need anything else from me.

grooverdan commented:
Despite the maintainers can edit this PR being checked I couldn't update the branch to enable the merge.
![image](https://github.com/user-attachments/assets/afda7527-9804-4744-8baf-1f7073bf6e98)

Can you do an update please.

oceanli-hub commented:
Done, I rebased the commit onto main again
